server/src/test/kotlin/io/spine/server/procman/given/pm/QuizProcess.kt

```diff
@@ -0,0 +1,176 @@
+        pmQuestionAnswered {
+            quiz = command.quiz
+            answer = command.answer
+        }
+
+    @React
+    fun on(event: PmQuizStarted): Nothing {
+        alter { id = event.quiz }
+        return nothing()
+    }
+
+    @React
+    fun on(event: PmQuestionAnswered): EitherOf3<PmQuestionSolved, PmQuestionFailed, Nothing> {
+        val answer = event.answer
+        val question = answer.question
+        if (question.isClosed()) {
+            return EitherOf3.withC(nothing())
+        }
+        return if (answer.correct) {
+            EitherOf3.withA(pmQuestionSolved {
+                quiz = event.quiz
+                this.question = question
+            })
+        } else {
+            EitherOf3.withB(pmQuestionFailed {
+                quiz = event.quiz
+                this.question = question
+            })
+        }
+    }
+
+    private fun PmQuestionId.isClosed(): Boolean {
+        val openQuestions = builder().openQuestionList
+        val isOpen = openQuestions.contains(this)
+        return !isOpen
+    }
+
+    @React
+    fun on(event: PmQuestionSolved): EitherOf2<Nothing, PmQuizFinished> {
+        val question = event.question
+        alter {
+            removeOpenQuestion(question)
+            addSolvedQuestion(question)
+        }
+        return onAnsweredQuestion()
+    }
+
+    @React
+    fun on(event: PmQuestionFailed): EitherOf2<Nothing, PmQuizFinished> {
+        val question = event.question
+        alter {
+            removeOpenQuestion(question)
+            addFailedQuestion(question)
+        }
+        return onAnsweredQuestion()
+    }
+
+    private fun PmQuiz.Builder.removeOpenQuestion(question: PmQuestionId) {
+        val openQuestions = openQuestionList
+        val index = openQuestions.indexOf(question)
+        removeOpenQuestion(index)
+    }
+
+    private fun onAnsweredQuestion(): EitherOf2<Nothing, PmQuizFinished> {
+        return if (builder().openQuestionList.isEmpty()) {
+            val loaded = select(PmQuizStats::class.java).findById(id())
+            EitherOf2.withB(pmQuizFinished {
+                quiz = id()
+                loaded?.let {
+                    stats = it
+                }
+            })
+        } else {
+            EitherOf2.withA(nothing())
+        }
+    }
+}
+
+internal class QuizRepository : ProcessManagerRepository<PmQuizId, QuizProcess, PmQuiz>()
+
+internal class QuizStatsView: Projection<PmQuizId, PmQuizStats, PmQuizStats.Builder>() {
+
+    @Subscribe
+    fun on(event: PmQuizStarted) = alter {
+        id = event.quiz
+    }
+
+    @Subscribe
+    @Suppress("UNUSED_PARAMETER")
+    fun on(event: PmQuestionAnswered) = alter {
+        solvedQuestions += 1
+    }
+
+    @Subscribe
+    @Suppress("UNUSED_PARAMETER")
+    fun on(event: PmQuestionFailed) = alter {
+        failedQuestions += 1
+    }
+}
+
+internal class QuizStatsRepository : ProjectionRepository<PmQuizId, QuizStatsView, PmQuizStats>()
```

Comment:
Same as above.

Reply:
Good point. I was keeping them for the case of custom routing. But it's not needed for now.